lightning/src/onion_message/messenger.rs

```diff
@@ -243,6 +244,89 @@ impl OnionMessageRecipient {
 	}
 }
 
+pub trait RespondFunction<T: OnionMessageContents>{
+	fn respond(&self, response: T, reply_path: BlindedPath);
+}
+
+impl<F, T> RespondFunction<T> for F
+where
+T: OnionMessageContents,
+F: Fn(T, BlindedPath)
+{
+	fn respond(&self, response: T, reply_path: BlindedPath) {
+		self(response, reply_path)
+	}
+}
+
+/// A struct handling response to an [`OnionMessage`]
+pub struct Responder<R, T>
+where
+R: RespondFunction<T>,
+T: OnionMessageContents
+{
+	messenger_function: R,
+	pub reply_path: BlindedPath,
+	// This phantom Data is used to ensure that we use T in the struct definition
+	// This allow us to ensure at compile time that the received message type and response type will be same
+	_phantom: PhantomData<T>
+}
+
+impl<R, T> Responder<R, T>
+where
+R: RespondFunction<T>,
+T: OnionMessageContents
+{
+    pub fn respond(&self, response: T) {
+		// Utilising the fact that we ensure at compile time that
+		// received message type, and response type will be same
+		self.messenger_function.respond(response, self.reply_path.clone())
+	}
+}
+
+/// A enum to handle received [`OnionMessage`]
+pub enum ReceivedOnionMessage<R, T>
+where
+R: RespondFunction<T>,
+T: OnionMessageContents
+{
+	WithReplyPath {
+		responder: Responder<R, T>,
```

Comment:
nit: let's make this the last field. Might be some sort of compiler optimizations if the first two fields in each variant are the same.

Reply:
Addressed in [2806122](https://github.com/lightningdevkit/rust-lightning/pull/2907/commits/2806122f65d57d3a6a91591aff6e9ccb7590af13)
Thanks!